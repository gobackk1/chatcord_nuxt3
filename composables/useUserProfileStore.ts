import { reactive } from 'vue'
import type firebase from 'firebase'
import admin from 'firebase-admin'
import useFirebase from './firebase/useFirebase'

export interface UserProfileStore {
  state: UserProfileState
  getters: UserProfileGetters
  actions: UserProfileActions
}

export interface UserProfileState {
  userData: CC.User | null
  friends: CC.P.Friend[]
  unsubscribeFriends: null | firebase.Unsubscribe
  friendRequests: CC.P.FriendRequest[]
  unsubscribeFriendRequests: null | firebase.Unsubscribe
}

export interface UserProfileGetters {
  getFriendByUid: (uid: string) => CC.P.Friend | null
}

export interface UserProfileActions {
  setUserData: (
    user: admin.auth.UserRecord | firebase.User | CC.User | null
  ) => void
  fetchFriends: (uid: string) => Promise<void>
  unsubscribeFriends: () => void
  fetchFriendRequests: (uid: string) => Promise<void>
  unsubscribeFriendRequests: () => void
  initializeStore: () => void
}

const { firestore, functions } = useFirebase()

const store: UserProfileStore = reactive<UserProfileStore>({
  state: {
    userData: null,
    friends: [],
    friendRequests: [],
    unsubscribeFriends: null,
    unsubscribeFriendRequests: null,
  },
  getters: {
    getFriendByUid(uid): CC.P.Friend | null {
      const target = store.state.friends.find((friend) => friend.uid === uid)
      return target === undefined ? null : target
    },
  },
  actions: {
    initializeStore() {
      store.state.userData = null

      store.state.friends = []

      store.state.friendRequests = []

      store.state.unsubscribeFriends?.()
      store.state.unsubscribeFriends = null

      store.state.unsubscribeFriendRequests?.()
      store.state.unsubscribeFriendRequests = null
    },
    setUserData(user) {
      if (user) {
        store.state.userData = {
          uid: user.uid,
          displayName: user.displayName,
          photoURL: user.photoURL,
        }
      } else {
        store.state.userData = null
      }
    },
    async fetchFriends(uid) {
      store.state.unsubscribeFriends = await firestore.fetchFriends({
        uid,
        handlers: {
          added: {
            handler: async (doc) => {
              const response = await functions.getUserByUid(doc.id)
              const { displayName, photoURL } = response.data.payload

              store.state.friends.push({
                displayName: displayName === null ? '' : displayName,
                photoURL: photoURL === null ? '' : photoURL,
                ...(doc.data() as CC.FS.Friend),
              })
            },
          },
          removed: {
            handler: (doc) => {
              const filteredFriends = store.state.friends.filter(
                (friend) => friend.uid !== doc.id
              )
              store.state.friends = filteredFriends
            },
          },
          completed: {
            handler: () => {
              // resolve()
              // return
            },
          },
        },
      })
    },
    unsubscribeFriends() {
      store.state.unsubscribeFriends?.()
    },
    async fetchFriendRequests(uid: string) {
      store.state.unsubscribeFriendRequests =
        await firestore.fetchFriendRequests({
          uid,
          handlers: {
            added: {
              handler: async (doc) => {
                const response = await functions.getUserByUid(doc.id)
                const { displayName, photoURL, uid } = response.data.payload
                store.state.friendRequests.push({
                  displayName: displayName === null ? '' : displayName,
                  photoURL: photoURL === null ? '' : photoURL,
                  uid,
                  ...(doc.data() as CC.FS.FriendRequest),
                })
              },
            },
            removed: {
              handler: (doc) => {
                const filteredFriendRequests =
                  store.state.friendRequests.filter(
                    (request) => request.uid !== doc.id
                  )
                store.state.friendRequests = filteredFriendRequests
              },
            },
            completed: {
              handler: () => {
                // resolve(unsubscribe)
              },
            },
          },
        })
    },
    unsubscribeFriendRequests() {
      store.state.unsubscribeFriendRequests?.()
    },
  },
})

export default function (): UserProfileStore {
  return store
}
