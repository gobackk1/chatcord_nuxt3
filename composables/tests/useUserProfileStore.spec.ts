import useUserProfileStore from '@/composables/useUserProfileStore'

const USER_RECORD = {
  uid: 'test',
  displayName: 'test user',
  photoURL: '/path/to/image.png',
}

describe('useUserProfileStoreのテスト', async () => {
  afterEach(() => {
    vi.restoreAllMocks()
  })
  describe('Actions', () => {
    describe('setUserData', () => {
      it('ユーザーデータを入力した場合、state.userDataが入力値になること', () => {
        const { state, actions } = useUserProfileStore()
        actions.setUserData(USER_RECORD)
        expect(state.userData).toStrictEqual(USER_RECORD)
        expect(state).toMatchSnapshot()
      })
      it('nullを入力した場合、state.userDataがnullになること', () => {
        const { state, actions } = useUserProfileStore()
        actions.setUserData(null)
        expect(state.userData).toBe(null)
        expect(state).toMatchSnapshot()
      })
    })
    describe('unsubscribeFriends', () => {
      it('購読解除用のメソッドを実行すること', () => {
        const { state, actions } = useUserProfileStore()
        const mock = vi.fn()
        state.unsubscribeFriends = mock
        actions.unsubscribeFriends()
        expect(mock).toHaveBeenCalledTimes(1)
        expect(state).toMatchSnapshot()
      })
    })
    describe('unsubscribeFriendRequests', () => {
      it('購読解除用のメソッドを実行すること', () => {
        const { state, actions } = useUserProfileStore()
        const mock = vi.fn()
        state.unsubscribeFriendRequests = mock
        actions.unsubscribeFriendRequests()
        expect(mock).toHaveBeenCalledTimes(1)
        expect(state).toMatchSnapshot()
      })
    })
    describe('initializeStore', () => {
      it('storeデータを初期化すること', () => {
        const { state, actions } = useUserProfileStore()
        const mockUnsubscribeFriends = vi.fn()
        const mockUnsubscribeFriendRequests = vi.fn()
        const fakeFriend = {
          ...USER_RECORD,
          dmId: 'test id',
        }
        const fakeFriendRequest = {
          ...USER_RECORD,
          type: 'send' as 'send' | 'receive',
        }
        state.userData = USER_RECORD
        state.friends = [fakeFriend]
        state.friendRequests = [fakeFriendRequest]
        state.unsubscribeFriends = mockUnsubscribeFriends
        state.unsubscribeFriendRequests = mockUnsubscribeFriendRequests

        actions.initializeStore()
        actions.unsubscribeFriends()
        actions.unsubscribeFriendRequests()
        expect(state.userData).toBe(null)
        expect(state.friends).toStrictEqual([])
        expect(state.friendRequests).toStrictEqual([])
        expect(state.unsubscribeFriends).toBe(null)
        expect(state.unsubscribeFriendRequests).toBe(null)
        expect(mockUnsubscribeFriends).toHaveBeenCalledTimes(1)
        expect(mockUnsubscribeFriendRequests).toHaveBeenCalledTimes(1)
        expect(state).toMatchSnapshot()
      })
    })
    describe.skip('fetchFriends')
    describe.skip('fetchFriendRequests')
  })
  describe('Getters', () => {
    describe('getFriendByUid', () => {
      it('入力したuidと同じuidを持つフレンドが存在する場合、フレンドのデータを返すこと', () => {
        const { state, getters } = useUserProfileStore()
        state.friends = [{ ...USER_RECORD, uid: 'frienduid', dmId: 'test id' }]
        const friend = getters.getFriendByUid('frienduid')
        expect(friend).toBeTruthy()
      })
      it('入力したuidと同じuidを持つフレンドが存在しない場合、nullを返すこと', () => {
        const { state, getters } = useUserProfileStore()
        state.friends = [{ ...USER_RECORD, uid: 'frienduid', dmId: 'test id' }]
        const friend = getters.getFriendByUid('nonexistentid')
        expect(friend).toBe(null)
      })
    })
  })
})
