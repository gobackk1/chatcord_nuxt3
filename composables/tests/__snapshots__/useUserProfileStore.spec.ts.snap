// Vitest Snapshot v1

exports[`useUserProfileStoreのテスト > Actions > initializeStore > storeデータを初期化すること 1`] = `
{
  "friendRequests": [],
  "friends": [],
  "unsubscribeFriendRequests": null,
  "unsubscribeFriends": null,
  "userData": null,
}
`;

exports[`useUserProfileStoreのテスト > Actions > setUserData > nullを入力した場合、state.userDataがnullになること 1`] = `
{
  "friendRequests": [],
  "friends": [],
  "unsubscribeFriendRequests": null,
  "unsubscribeFriends": null,
  "userData": null,
}
`;

exports[`useUserProfileStoreのテスト > Actions > setUserData > ユーザーデータを入力した場合、state.userDataが入力値になること 1`] = `
{
  "friendRequests": [],
  "friends": [],
  "unsubscribeFriendRequests": null,
  "unsubscribeFriends": null,
  "userData": {
    "displayName": "test user",
    "photoURL": "/path/to/image.png",
    "uid": "test",
  },
}
`;

exports[`useUserProfileStoreのテスト > Actions > unsubscribeFriendRequests > 購読解除用のメソッドを実行すること 1`] = `
{
  "friendRequests": [],
  "friends": [],
  "unsubscribeFriendRequests": [MockFunction spy] {
    "calls": [
      [],
    ],
    "results": [
      {
        "type": "return",
        "value": undefined,
      },
    ],
  },
  "unsubscribeFriends": [MockFunction spy],
  "userData": null,
}
`;

exports[`useUserProfileStoreのテスト > Actions > unsubscribeFriends > 購読解除用のメソッドを実行すること 1`] = `
{
  "friendRequests": [],
  "friends": [],
  "unsubscribeFriendRequests": null,
  "unsubscribeFriends": [MockFunction spy] {
    "calls": [
      [],
    ],
    "results": [
      {
        "type": "return",
        "value": undefined,
      },
    ],
  },
  "userData": null,
}
`;
